### PR TITLE
Elasticsearch: Add generic support for template variables

### LINKED ISSUE #7794: Elasticsearch: Support template variables as the Size of Group by Terms
Currently the template variable is parsed as an int and I'm getting ```illegal_argument_exception```: ```[terms] size doesn't support values of type: VALUE_NULL```

group_by__size

Grafana v4.3.0-pre1 (commit: 2f5814d51)
ES 5.x
Windows 7 x64

### LINKED ISSUE #8339: ElasticSearch: Permit "field" to be parsed to a variable
Grafana version: 4.3.0-pre1
Datasource: Elasticsearch
OS: CentOS 7
If is tried to use "field": "$variable" in query of Variables, the variable is not resolved and the query for the elasticsearch.

Tryied:
Create a variable named $test, type Custom with value "AAA,BBB"
Select $test ='AAA' in a dashboard.
Set another variable named named $test1 with query = {"find": "terms", "field": "$test"}

Result: 
{"find": "terms", "field": "$test"}

Espected result:
{"find": "terms", "field": "AAA"}

### LINKED ISSUE #24239: Elasticsearch alias with template variables not working as expected.
**What happened**:
When using templates variables to select the group by term and using an alias like {{term $variable}}, all the series will have the alias 'undefined'
**What you expected to happen**:
The alias will use the value of the selected group by term. 

**How to reproduce it (as minimally and precisely as possible)**:
On the templated Elasticsearch playground [1], creating a custom variable with the value `@hostname` and using it both in the terms group by and in the alias.
image
NOTE:  The error is only present when the template variable is used in the group by. (if I use @hostname as the term and use the variable in the alias, it works as expected).

**Anything else we need to know?**:

**Environment**:
- Grafana version: tested on 6.7.3 and 7.0.0-beta1
- Data source type & version: Elasticsearch 6.x, 7.x
- OS Grafana is installed on: Centos 7.x
- User OS & Browser: All. 
- Grafana plugins: All. 
[1] https://play.grafana.org/d/000000015/elasticsearch-templated?editPanel=1&orgId=1

### PR DESCRIPTION
**Pull request by Elfo404**
**What this PR does / why we need it**:
Another shot at https://github.com/grafana/grafana/pull/29739 (adding generic support for template variables) using a different approach as suggested by @torkelo. Seems to work as expected, but I want to do some more manual testing to be sure I'm not breaking the hell out of Elasticsearch.

The code in the query builder is really untyped, but this should be a minor issue given we are moving to a BE-only query builder.

**Which issue(s) this PR fixes**:
Fixes #7794
Fixes #8339
Fixes #24239
Closes #29739

**Special notes for your reviewer**:

There are a few places where we interpolate variables other than at the beginning of the `query` method. I'm pretty sure it should be safe to remove them, but I'll leave it for a follow up PR.

/cc. @simianhacker @estermv

### CODE REVIEW

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/MovingAverageSettingsEditor.tsx [hunk 1]
```diff
@@ -1,159 +1,98 @@
...
-            />
-          </InlineField>
-          <InlineField label="Gamma">
-            <Input
-              onBlur={(e) =>
-                dispatch(
-                  changeMetricSetting(metric, 'settings', {
-                    ...metric.settings?.settings,
-                    gamma: parseInt(e.target.value!, 10),
+                    gamma: e.target.value!,
```
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/MovingAverageSettingsEditor.tsx:73]: note to self: remove `!`

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/aggregations.ts [hunk 2]
```diff
@@ -183,26 +183,28 @@ export interface MovingAverageLinearModelSettings extends BaseMovingAverageModel
 
 export interface MovingAverageEWMAModelSettings extends BaseMovingAverageModelSettings {
   model: 'ewma';
-  alpha: number;
+  settings?: {
+    alpha?: string;
+  };
```
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/aggregations.ts:188]: this wasn't actually working before

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/aggregations.ts [hunk 3]
```diff
@@ -235,6 +237,10 @@ export const isHoltWintersMovingAverage = (
   metric: MovingAverage | MovingAverage<'holt_winters'>
 ): metric is MovingAverage<'holt_winters'> => metric.settings?.model === 'holt_winters';
 
+export const isMovingAverageWithModelSettings = (
+  metric: MovingAverage
+): metric is MovingAverage<'ewma'> | MovingAverage<'holt'> | MovingAverage<'holt_winters'> =>
+  ['holt', 'ewma', 'holt_winters'].includes(metric.settings?.model || '');
```
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/aggregations.ts:243]: note to self: add a blank line

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/utils.ts [hunk 4]
```diff
@@ -160,7 +160,11 @@ export const metricAggregationConfig: MetricsConfiguration = {
     hasSettings: true,
     supportsInlineScript: false,
     hasMeta: false,
-    defaults: {},
+    defaults: {
+      settings: {
+        lag: '1',
```
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/utils.ts:165]: copying ES defaults to have more predictable default values

File: pkg/tsdb/elasticsearch/time_series_query.go [hunk 5]
```diff
@@ -179,6 +179,49 @@ func (e *timeSeriesQuery) processQuery(q *Query, ms *es.MultiSearchRequestBuilde
 	return nil
 }
 
+// Casts values to int when required by Elastic's query DSL
+func (metricAggregation MetricAgg) castSettings() map[string]interface{} {
```
Comment (Elfo404) [inline pkg/tsdb/elasticsearch/time_series_query.go:183]: Naming is hard.
This converts a query model settings to an ES DSL compatible object by casting required values without altering the source variable.

`getDSLCompatibleValues`, `getCastedValues`, `formatSettings`? 🤔

Comment (Elfo404) [reply pkg/tsdb/elasticsearch/time_series_query.go]: the current `generateSettingsForDSL` maybe?

File: pkg/tsdb/elasticsearch/time_series_query_test.go [hunk 6]
```diff
@@ -853,6 +853,80 @@ func TestExecuteTimeSeriesQuery(t *testing.T) {
 				"var1": "_count",
 			})
 		})
+
+		Convey("With aggregations that requires type casting", func() {
```
Comment (dsotirakis) [inline pkg/tsdb/elasticsearch/time_series_query_test.go:857]: There's a plan to move away from using `convey` and start using `testify`, asserting with `assert.` or `require.`.

Shouldn't be much trouble to do it - [check this out](https://github.com/grafana/grafana/blob/master/pkg/services/alerting/rule_test.go)

**Review by dsotirakis — COMMENTED** [review]
Left a small comment!

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 7]
```diff
@@ -353,26 +356,39 @@ export class ElasticDatasource extends DataSourceApi<ElasticsearchQuery, Elastic
   }
 
   interpolateVariablesInQueries(queries: ElasticsearchQuery[], scopedVars: ScopedVars): ElasticsearchQuery[] {
```
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:358]: I didn't add tests for this as it is already tested and *should* have a good coverage

File: pkg/tsdb/elasticsearch/time_series_query.go [hunk 8]
```diff
@@ -179,6 +179,30 @@ func (e *timeSeriesQuery) processQuery(q *Query, ms *es.MultiSearchRequestBuilde
 	return nil
 }
 
+// Casts values to int when required by Elastic's query DSL
+func (metricAggregation MetricAgg) generateSettingsForDSL() map[string]interface{} {
+	setIntPath := func(path ...string) {
+		if stringValue, err := metricAggregation.Settings.GetPath(path...).String(); err == nil {
+			value, _ := strconv.Atoi(stringValue)
```
Comment (dsotirakis) [inline pkg/tsdb/elasticsearch/time_series_query.go:186]: Was thinking about:

```suggestion
			if value, err := strconv.Atoi(stringValue); err == nil {
			metricAggregation.Settings.SetPath(path, value)
		}
```

Comment (dsotirakis) [reply pkg/tsdb/elasticsearch/time_series_query.go:186]: What do you think?

Comment (Elfo404) [reply pkg/tsdb/elasticsearch/time_series_query.go:186]: yup! I'll update! 👍

File: pkg/tsdb/elasticsearch/time_series_query.go [hunk 9]
```diff
@@ -179,6 +179,31 @@ func (e *timeSeriesQuery) processQuery(q *Query, ms *es.MultiSearchRequestBuilde
...
+				metricAggregation.Settings.SetPath(path, value)
+			}
+		}
+	}
+
+	switch metricAggregation.Type {
+	case "moving_avg":
+		setIntPath("window")
+		setIntPath("predict")
+		setIntPath("settings", "alpha")
```
Comment (ifrost) [inline pkg/tsdb/elasticsearch/time_series_query.go:196]: Why alpha, beta, gamma is parse to int? To should be a float value 0-1

File: public/app/plugins/datasource/elasticsearch/query_builder.ts [hunk 10]
```diff
@@ -355,6 +387,15 @@ export class ElasticQueryBuilder {
     return query;
   }
 
+  private toNumber(stringValue: unknown): unknown | number {
+    const parsedValue = parseInt(`${stringValue}`, 10);
```
Comment (ifrost) [inline public/app/plugins/datasource/elasticsearch/query_builder.ts:391]: Similar questions as for alpha, beta, gamma settings in the backend - they should be floats 0-1 so I guess we shouldn't parse them to ints?

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/MovingAverageSettingsEditor.tsx [hunk 11]
```diff
@@ -1,159 +1,98 @@
...
-            <Switch
-              onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
-                dispatch(changeMetricSetting(metric, 'minimize', e.target.checked))
-              }
-              checked={!!metric.settings?.minimize}
-            />
-          </InlineField>
-        </>
+      {(isEWMAMovingAverage(metric) || isHoltMovingAverage(metric) || isHoltWintersMovingAverage(metric)) && (
+        <InlineField label="Alpha" labelWidth={16}>
```
Comment (ifrost) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/MovingAverageSettingsEditor.tsx:34]: Could we use `SettingField` here? (and for other fields that are using `InlineField`)

Comment (ifrost) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/MovingAverageSettingsEditor.tsx:34]: I guess we can't coz it requires entire settings object 🤔 

And Now for Something Completely Different: `alpha` prop is moving from `metric` to `metric.settings` for EWMA. It means that whatever user had saved in alpha will be gone after this change. Is it something worth mentioning as a breaking change?

File: pkg/tsdb/elasticsearch/time_series_query.go (hunk 9 above)
Comment (ifrost) [reply pkg/tsdb/elasticsearch/time_series_query.go:196]: Oh, just checked that it's the current behaviour in master (for dashboard queries and alerts), but isn't it a bug that could be addressed separately?

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 12]
```diff
@@ -353,26 +356,39 @@ export class ElasticDatasource extends DataSourceApi<ElasticsearchQuery, Elastic
...
+    const expandedQueries = queries.map(
+      (query): ElasticsearchQuery => ({
+        ...query,
+        datasource: this.name,
+        query: this.interpolateLuceneQuery(query.query || '', scopedVars),
+        bucketAggs: query.bucketAggs?.map(interpolateBucketAgg),
+      })
+    );
+
+    const finalQueries: ElasticsearchQuery[] = JSON.parse(
```
Comment (ifrost) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:387]: Why do we need `interpolateBucketAgg`? Wouldn't `this.templateSrv.replace(JSON.stringify(expandedQueries), scopedVars)` interpolate it correctly?

Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:387]: because if it a bucketAgg is `filters` we need to use `interpolateLuceneQuery` in each `settings.filters[n].query` property. 

`interpolateBucketAgg` is a noop for all other aggregations

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/MovingAverageSettingsEditor.tsx (hunk 11 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/SettingsEditor/MovingAverageSettingsEditor.tsx:34]: > I guess we can't coz it requires entire settings object

yeah, that's eventually something I want to change as I don't really like it as it is

File: pkg/tsdb/elasticsearch/time_series_query.go (hunk 9 above)
Comment (Elfo404) [reply pkg/tsdb/elasticsearch/time_series_query.go:196]: that's true, i actually overlooked it! good catch!

Comment (Elfo404) [reply pkg/tsdb/elasticsearch/time_series_query.go:196]: actually everything should be a float, not an int.

Comment (Elfo404) [reply pkg/tsdb/elasticsearch/time_series_query.go:196]: > Also, how can we ensure that every time a new field is added this list is updated too?

my idea on the backend is that every type of aggregation will have a MarshalJson function, so that each of them can define only for itself what should be its own Elasticsearch DSL JSON representation. this is a short-term solution to make what we already have working.
makes sense?

File: public/app/plugins/datasource/elasticsearch/query_builder.ts (hunk 10 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/query_builder.ts]: updated 👍

File: pkg/tsdb/elasticsearch/time_series_query.go (hunk 9 above)
Comment (ifrost) [reply pkg/tsdb/elasticsearch/time_series_query.go]: Coolio 👍 Makes sense. Before that happens - would it be worth leaving a comment with cross-reference between `public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/aggregations.ts` <->  `pkg/tsdb/elasticsearch/time_series_query.go:181` <-> `public/app/plugins/datasource/elasticsearch/query_builder.ts:344` as a reminder that if one is changed the others should be updated too?

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 12 above)
Comment (ifrost) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:387]: Oh I got it now. It's because we need to default to `*`. It's actually covered in tests but it's kind of tricky to see it straight away. IMO it'd be a bit easier if it was more an integration test using the real `TemplateSrv` and an extra case to explicitly pass vars that have empty values - something like this:
https://github.com/grafana/grafana/commit/229d0c8a82f9f6dafe224c03e496ba7be70f4a74 
But yeah, the coverage should be the same anyway so it's up to you 👍

**Review by ifrost — APPROVED** [review]

**Comment by ifrost** [comment]
clap

File: public/app/plugins/datasource/elasticsearch/datasource.ts (hunk 12 above)
Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:387]: actually `interpolateLuceneQuery` uses a different interpolation format (https://grafana.com/docs/grafana/latest/variables/advanced-variable-format-options/#lucene---elasticsearch) AND default to `*` if the interpolation results in empty strings.

> and an extra case to explicitly pass vars that have empty values

you mean something like: 
https://github.com/grafana/grafana/blob/bad048b7ba1fecd28e76509c545a67fc8b70fdeb/public/app/plugins/datasource/elasticsearch/datasource.test.ts#L880

?

Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:387]: > vars that have empty values

aaaaaah, I got your point now, lemme check. 👍

Comment (Elfo404) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:387]: i took a better look at it and it should be already covered, I don't feel it will ad much value adding it here, so I'm gonna merge it :)
